_posts/2023-12-31-end-of-year-update.markdown

```diff
@@ -0,0 +1,145 @@
+
+### ABI switch
+_Written by @Dennisbonke_
+
+mlibc has two big ABIs that a port can use, these being the [mlibc ABI](https://github.com/managarm/mlibc/tree/master/abis/mlibc) and the [Linux ABI](https://github.com/managarm/mlibc/tree/master/abis/linux). Managarm has from the very start always used the mlibc ABI, but some software expects the Linux ABI, by hardcoding it or asserting it at compile time. So we decided to move Managarm to the Linux ABI, which involved replacing some hardcoded constants in the kernel with their actual defines and recompiling the entire world against this new ABI. Not neccesarily the most shiny work, but useful none the less.
+
+### Support for C11 Threads in mlibc
+_Written by @no92_
+
+Porting [`foot`](https://codeberg.org/dnkl/foot/) revealed that mlibc didn't support C11 threads yet. But as it turns out, they are really similar to how pthreads work. This made the implementation easy - the code was adapted to be generic to the two interfaces and moved into mlibc internals - now C11 threads and pthreads mostly just call those with the correct parameters.
+
+### Changes in the DRM subsystem
+
+While working on GPU support it became apparent that eir, our pre-kernel, doesn't handle LFB addresses over 4 GiB well, as it is limited to 32-bit addresses. The [workaround](https://github.com/managarm/managarm/commit/9db754ded8d0fd3efffe156c899b9e6868010448) is to ignore the framebuffer in eir if that happens, as thor will pick it up and use it correctly regardless. As we had not previously encountered such a case before, it might be limited to GPU passthrough scenarios.
+
+The virtio-gpu driver received a cleanup this year. As the mesa support can be enabled at configure time, it would be interesting to implement the OS-side for 3D acceleration - a project for next year, maybe.
+
+## New ports and updates
+
+### Mesa upstreaming
+_Written by @ElectrodeYT_
+
+In our quest to reduce the amount of patches needed for packages, we also started to look at patches we could easily upstream. One of the ones that jumped out was mesa: it is a core package, part of `weston-desktop`, and had patches which (after removal of some no longer needed patches) were extremely simple, and basically amounted to some `#ifdef`s and some minor changes to `meson.build`.
+
+Upstreaming it went smoothly, however it was not done in time to get into version 23.3 of mesa, meaning that, for the time being, patches are still necessary, however this will soon enough change.
```

Comment:
```suggestion
Upstreaming it went smoothly, however it was not done in time to get into version 23.3 of Mesa, meaning that, for the time being, patches are still necessary, however this will soon enough change.
```